README.md
# SwiftSpinner

[![Version](https://img.shields.io/cocoapods/v/SwiftSpinner.svg?style=flat)](http://cocoadocs.org/docsets/SwiftSpinner)
[![License](https://img.shields.io/cocoapods/l/SwiftSpinner.svg?style=flat)](http://cocoadocs.org/docsets/SwiftSpinner)
[![Platform](https://img.shields.io/cocoapods/p/SwiftSpinner.svg?style=flat)](http://cocoadocs.org/docsets/SwiftSpinner)
![Swift 5.0](https://img.shields.io/badge/swift-5.0-orange.svg)

SwiftSpinner is an extra beautiful activity indicator with plain and bold style. It uses dynamic blur and translucency to overlay the current screen contents and display an activity indicator with text (or the so called “spinner”).

- - -

I developed it for my Swift app called **Doodle Doodle** and wanted to share it with everyone.

This is how the activity looks like (from the demo app):

![SwiftSpinner Screenshot](https://raw.githubusercontent.com/icanzilb/SwiftSpinner/main/etc/spinner-preview.gif)

## Usage

To run the example project, clone the repo, and run `pod install` from the DemoApp directory first. That’ll run the demo program which shows you how the spinner looks like and what it can do.

#### Code samples

The simple code to get `SwiftSpinner` running in your own app.

 * In case you installed SwiftSpinner via CocoaPods you need to import it (add this somewhere at the top of your source code file):

```swift
import SwiftSpinner
```

 * When you want to show an animated activity (eg. rings are randomly rotating around):

```swift
SwiftSpinner.show("Connecting to satellite...")
```

 * If you want to show a static activity indicator (eg. a message with two complete rings around it)

```swift
SwiftSpinner.show("Failed to connect, waiting...", animated: false)
```

 * When you want to hide the activity:

```swift
SwiftSpinner.hide()
```

In case you want to do something after the hiding animation completes you can provide a closure to the `hide()` method:

```swift
SwiftSpinner.hide({
  //do stuff
})
```


That's all. If you want to change the text of the current activity, just call `show(...)` again, this will animate the old text into the new text.

## Beyond the basics

If you are using `SwiftSpinner` to show an alert message you can also easily add a dismiss handler:

```swift
SwiftSpinner.show("Connecting \nto satellite...").addTapHandler({
  SwiftSpinner.hide()
})
```

Or even add a subtitle to let the user know they can tap to do stuff:

```swift
SwiftSpinner.show("Connecting \nto satellite...").addTapHandler({
  SwiftSpinner.hide()
}, subtitle: "Tap to hide while connecting! This will affect only the current operation.")
```

In case you want to adjust the font of the spinner title:

```swift
SwiftSpinner.setTitleFont(UIFont(name: "Futura", size: 22.0))
```

To reset back to the default font:

```swift
SwiftSpinner.setTitleFont(nil)
```

In case you want to change an arbitrary aspect of the text on screen access directly:

```swift
SwiftSpinner.shared.titleLabel
SwiftSpinner.shared.subtitleLabel
```

You can show a spinner only if certain amount of time has passed (e.g. if you are downloading a file - show a message only if the operation takes longer than certain amount of time):

```swift
SwiftSpinner.show(delay: 2.0, title: "It's taking longer than expected")
```

If you call `show(…)` or `hide()` before the `delay` time has passed - this will clear the call to `show(delay: …)`.

You show a message for a certain duration:
```swift
SwiftSpinner.show(duration: 4.0, title: "It's taking longer than expected")
```

Or you can use `SwiftSpinner` as a progress bar by directly setting the current progress like so:

```swift
SwiftSpinner.show(progress: 0.2, title: "Downloading Data...") // 20% trough the process
```

If you want to see the content behind the spinner, set the showBlurBackground to false:
```swift
SwiftSpinner.showBlurBackground = false
```

## Requirements

UIKit must be imported. If you are using SwiftSpinner in an App Extension, you must add `EXTENSION` to your `Other Swift Flags` Build Settings.

![Extension Setting Screenshot](https://user-images.githubusercontent.com/444725/26855417-a7a8241a-4acf-11e7-8528-b37a28448113.png)

## Installation

`SwiftSpinner` is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your `Podfile`:

```
pod "SwiftSpinner"
```

In case you don’t want to use CocoaPods - just copy the file **SwiftSpinner/SwiftSpinner.swift** to your Xcode project.

Credit
========

Author: **Marin Todorov**

* [http://www.underplot.com](http://www.underplot.com)
* [https://twitter.com/icanzilb](https://twitter.com/icanzilb)

More about Marin:

<table>
<tr>
<td>
<a href="http://www.ios-animations-by-tutorials.com/"><img src="http://www.underplot.com/images/thumbs/iat.jpg" width="170"><br>
<b>iOS Animations by Tutorials</b>, Author</a>
</td>
<td>
<a href="http://www.ios-animations-by-emails.com/"><img src="http://www.underplot.com/images/thumbs/ios-animations-by-emails.jpg" width="170"><br>
iOS Animations by Emails Newsletter, Author</a>
</td>
</tr>
</table>

## License

`SwiftSpinner` is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`SwiftSpinner` adheres to [Semantic Versioning](http://semver.org/).

- `2.2.x` Releases - [2.2.0](#220)
- `2.1.x` Releases - [2.1.0](#210)
- `2.0.x` Releases - [2.0.0](#200)
- `1.6.x` Releases - [1.6.0](#160)
- `1.5.x` Releases - [1.5.0](#150)
- `1.4.x` Releases - [1.4.0](#140)

---

## [2.2.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/2.2.0)

- Dropped iOS 8 compatibility 

## [2.1.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/2.1.0)

- SwiftPM compatibility  

## [2.0.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/2.0.0)

### Breaking change

The singleton property has been renamed to `shared` in accordance with other platform conventions.

### Other changes

- Added a completion handler to the `hdie` method, see #135
- Merged #136
- Merged #134

## [1.6.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/1.6.0)

- Support Swift 4.2
- Merge #122
- Merge #127

## [1.5.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/1.5.0)

- hierarchy edge case fix
- add more color customization options

## [1.4.0](https://github.com/icanzilb/SwiftSpinner/releases/tag/1.4.0)

- Support Swift 4


DemoApp/SwiftSpinnerDemo/AppDelegate.swift
//
// Copyright (c) 2015 Marin Todorov, Underplot ltd.
// This code is distributed under the terms and conditions of the MIT license.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

  var window: UIWindow?

  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    // Override point for customization after application launch.
    return true
  }

}


DemoApp/SwiftSpinnerDemo/ViewController.swift
//
// Copyright (c) 2015-present Marin Todorov, Underplot ltd.
// This code is distributed under the terms and conditions of the MIT license.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import UIKit
import SwiftSpinner

class ViewController: UIViewController {
  
    @IBOutlet var labelContainerView: UIView!
    var progress = 0.0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        labelContainerView.transform = CGAffineTransform(translationX: 0, y: 200)
    }
    
    func delay(seconds: Double, completion: @escaping () -> ()) {
        let popTime = DispatchTime.now() + Double(Int64( Double(NSEC_PER_SEC) * seconds )) / Double(NSEC_PER_SEC)
        
        DispatchQueue.main.asyncAfter(deadline: popTime) {
            completion()
        }
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        self.demoSpinner()
    }
    
    func demoSpinner() {
        UIView.animate(withDuration: 0.5) {
            self.labelContainerView.transform = .identity
        }

        SwiftSpinner.show(delay: 0.5, title: "Shouldn't see this one", animated: true)
        SwiftSpinner.hide()
        
        SwiftSpinner.show(delay: 1.0, title: "Connecting...", animated: true)
        
        delay(seconds: 2.0, completion: {
            SwiftSpinner.show("Connecting \nto satellite...").addTapHandler({
                print("tapped")
                SwiftSpinner.hide()
            }, subtitle: "Tap to hide while connecting! This will affect only the current operation.")
        })
        
        delay(seconds: 6.0, completion: {
            SwiftSpinner.show("Authenticating user account")
        })
        
        delay(seconds: 10.0, completion: {
            SwiftSpinner.shared.outerColor = UIColor.red.withAlphaComponent(0.5)
            SwiftSpinner.setTitleColor(UIColor.red)
            SwiftSpinner.show("Failed to connect, waiting...", animated: false)
        })
        
        delay(seconds: 14.0, completion: {
            SwiftSpinner.shared.outerColor = nil
            SwiftSpinner.setTitleFont(UIFont(name: "Futura", size: 22.0))
            SwiftSpinner.setTitleColor(UIColor.white)
            SwiftSpinner.show("Retrying to authenticate")
        })
        
        delay(seconds: 18.0, completion: {
            SwiftSpinner.setTitleColor(UIColor.green)
            SwiftSpinner.show("Connecting...")
        })
        
        delay(seconds: 21.0, completion: {
            SwiftSpinner.setTitleFont(nil)
            SwiftSpinner.shared.innerColor = UIColor.green.withAlphaComponent(0.5)
            SwiftSpinner.show(duration: 2.0, title: "Connected", animated: false)
        })

        delay(seconds: 24.0) {
            SwiftSpinner.setTitleColor(UIColor.white)
            SwiftSpinner.shared.innerColor = nil
            Timer.scheduledTimer(timeInterval: 0.1, target: self, selector: #selector(self.timerFire), userInfo: nil, repeats: true)
        }
        
        delay(seconds: 34.0, completion: {
            self.demoSpinner()
        })
    }
    
    @objc func timerFire(_ timer: Timer) {
        progress += (timer.timeInterval/5)
        SwiftSpinner.show(progress: progress, title: "Downloading: \(Int(progress * 100))% completed")
        if progress >= 1 {
            progress = 0
            timer.invalidate()
            SwiftSpinner.show(duration: 2.0, title: "Complete!", animated: false)
        }
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
}


DemoApp/SwiftSpinnerDemoTests/SwiftSpinnerDemoTests.swift
//
// Copyright (c) 2015 Marin Todorov, Underplot ltd.
// This code is distributed under the terms and conditions of the MIT license.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// The MIT License in plain English: http://www.touch-code-magazine.com/JSONModel/MITLicense

import UIKit
import XCTest

class SwiftSpinnerDemoTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Sources/SwiftSpinner/SwiftSpinner.swift
//
// Copyright (c) 2015-present Marin Todorov, Underplot ltd.
// This code is distributed under the terms and conditions of the MIT license.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import UIKit

public class SwiftSpinner: UIView {
    fileprivate static let standardAnimationDuration = 0.33

   // MARK: - Singleton

    //
    // Access the singleton instance
    //
    public static let shared = SwiftSpinner(frame: CGRect.zero)

    // MARK: - Init

    /// Init
    ///
    /// - Parameter frame: the view's frame
    public override init(frame: CGRect) {
        currentTitleFont = defaultTitleFont // By default we initialize to the same.

        super.init(frame: frame)

        blurEffect = UIBlurEffect(style: blurEffectStyle)
        blurView = UIVisualEffectView()
        addSubview(blurView)

        vibrancyView = UIVisualEffectView(effect: UIVibrancyEffect(blurEffect: blurEffect))
        addSubview(vibrancyView)

        let titleScale: CGFloat = 0.85
        titleLabel.frame.size = CGSize(width: frameSize.width * titleScale, height: frameSize.height * titleScale)
        titleLabel.font = currentTitleFont
        titleLabel.numberOfLines = 0
        titleLabel.textAlignment = .center
        titleLabel.lineBreakMode = .byWordWrapping
        titleLabel.adjustsFontSizeToFitWidth = true
        titleLabel.textColor = UIColor.white

        blurView.contentView.addSubview(titleLabel)
        blurView.contentView.addSubview(vibrancyView)

        outerCircleView.frame.size = frameSize

        outerCircle.path = UIBezierPath(ovalIn: CGRect(x: 0.0, y: 0.0, width: frameSize.width, height: frameSize.height)).cgPath
        outerCircle.lineWidth = 8.0
        outerCircle.strokeStart = 0.0
        outerCircle.strokeEnd = 0.45
        outerCircle.lineCap = .round
        outerCircle.fillColor = UIColor.clear.cgColor
        outerCircle.strokeColor = outerCircleDefaultColor
        outerCircleView.layer.addSublayer(outerCircle)

        outerCircle.strokeStart = 0.0
        outerCircle.strokeEnd = 1.0

        blurView.contentView.addSubview(outerCircleView)

        innerCircleView.frame.size = frameSize

        let innerCirclePadding: CGFloat = 12
        innerCircle.path = UIBezierPath(ovalIn: CGRect(x: innerCirclePadding, y: innerCirclePadding, width: frameSize.width - 2*innerCirclePadding, height: frameSize.height - 2*innerCirclePadding)).cgPath
        innerCircle.lineWidth = 4.0
        innerCircle.strokeStart = 0.5
        innerCircle.strokeEnd = 0.9
        innerCircle.lineCap = .round
        innerCircle.fillColor = UIColor.clear.cgColor
        innerCircle.strokeColor = innerCircleDefaultColor
        innerCircleView.layer.addSublayer(innerCircle)

        innerCircle.strokeStart = 0.0
        innerCircle.strokeEnd = 1.0

        blurView.contentView.addSubview(innerCircleView)

        isUserInteractionEnabled = true
    }

    public override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        return self
    }

    // MARK: - Public interface

    /// The label with the spinner's title
    public lazy var titleLabel = UILabel()

    /// The label with the spinner's subtitle
    public var subtitleLabel: UILabel?

    private let outerCircleDefaultColor = UIColor.white.cgColor
    fileprivate var _outerColor: UIColor?

    /// The color of the outer circle
    public var outerColor: UIColor? {
        get { return _outerColor }
        set(newColor) {
            _outerColor = newColor
            outerCircle.strokeColor = newColor?.cgColor ?? outerCircleDefaultColor
        }
    }

    private let innerCircleDefaultColor = UIColor.gray.cgColor
    fileprivate var _innerColor: UIColor?

    /// The color of the inner circle
    public var innerColor: UIColor? {
        get { return _innerColor }
        set(newColor) {
            _innerColor = newColor
            innerCircle.strokeColor = newColor?.cgColor ?? innerCircleDefaultColor
        }
    }

    /// Custom superview for the spinner
    private static weak var customSuperview: UIView?
    private static func containerView() -> UIView? {
        #if EXTENSION
            return customSuperview
        #else
            if #available(iOS 13, *) {
                return customSuperview ?? UIApplication.shared.windows.first{ $0.isKeyWindow }
            } else {
                return customSuperview ?? UIApplication.shared.keyWindow
            }
        #endif
    }

    /// Custom container for the spinner
    public class func useContainerView(_ sv: UIView?) {
        customSuperview = sv
    }

    /// Show the blurred background. If false the background content will be visible. Defaults to true.
    public static var showBlurBackground: Bool = true

    /// Show the spinner activity on screen, if visible only update the title
    ///
    /// - Parameters:
    ///   - title: The title shown under the spiiner
    ///   - animated: Animate the spinner. Defaults to true
    /// - Returns: The instance of the spinner
    @discardableResult
    public class func show(_ title: String, animated: Bool = true) -> SwiftSpinner {
        let spinner = SwiftSpinner.shared

        spinner.clearTapHandler()

        spinner.updateFrame()

        if spinner.superview == nil {
            // Show the spinner
            spinner.blurView.contentView.alpha = 0

            guard let containerView = containerView() else {
                #if EXTENSION
                    fatalError("\n`containerView` is `nil`. `UIApplication.keyWindow` is not available in extensions and so, a containerView is required. Use `useContainerView` to set a view where the spinner should show")
                #else
                    fatalError("\n`UIApplication.keyWindow` is `nil`. If you're trying to show a spinner from your view controller's `viewDidLoad` method, do that from `viewWillAppear` instead. Alternatively use `useContainerView` to set a view where the spinner should show")
                #endif
            }

            containerView.addSubview(spinner)

            UIView.animate(withDuration: SwiftSpinner.standardAnimationDuration, delay: 0.0, options: .curveEaseOut, animations: {
                spinner.blurView.contentView.alpha = 1
                spinner.blurView.effect = showBlurBackground ? spinner.blurEffect : .none
            }, completion: nil)

            #if os(iOS)
                // Orientation change observer
                NotificationCenter.default.addObserver(
                    spinner,
                    selector: #selector(SwiftSpinner.orientationChangedAction),
                    name: UIDevice.orientationDidChangeNotification,
                    object: nil)
            #endif
        } else if spinner.dismissing {
            // If the spinner is hiding, delay the next show. The duration is set to double the standard animation to avoid an edge case that caused endless laoding. See #125
            show(delay: SwiftSpinner.standardAnimationDuration, title: title, animated: true)
        }

        spinner.title = title
        spinner.animating = animated

        return spinner
    }

    /// Show the spinner activity on screen with duration, if visible only update the title
    ///
    /// - Parameters:
    ///   - duration: The duration of the show animation
    ///   - title: The title shown under the spinner
    ///   - animated: Animate the spinner. Defaults to true
    ///   - completion: An optional completion handler
    /// - Returns: The instance of the spinner
    @discardableResult
    public class func show(duration: Double, title: String, animated: Bool = true, completion: (() -> ())? = nil) -> SwiftSpinner {
        let spinner = SwiftSpinner.show(title, animated: animated)
        spinner.delay(duration) {
            SwiftSpinner.hide {
                completion?()
            }
        }
        return spinner
    }

    private static var delayedTokens = [String]()

    /// Show the spinner activity on screen, after delay. If new call to show, showWithDelay or hide is maked before execution this call is discarded
    ///
    /// - Parameters:
    ///   - delay: The delay time
    ///   - title: The title shown under the spinner
    ///   - animated: Animate the spinner. Defaults to true
    public class func show(delay: Double, title: String, animated: Bool = true) {
        let token = UUID().uuidString
        delayedTokens.append(token)
        SwiftSpinner.shared.delay(delay, completion: {
            if let index = delayedTokens.firstIndex(of: token) {
                delayedTokens.remove(at: index)
                SwiftSpinner.show(title, animated: animated)
            }
        })
    }

    /// Show the spinner with the outer circle representing progress (0 to 1)
    ///
    /// - Parameters:
    ///   - progress: The progress percentage. Values between 0 and 1
    ///   - title: The title shown under the spinner
    /// - Returns: The instance of the spinner
    @discardableResult
    public class func show(progress: Double, title: String) -> SwiftSpinner {
        let spinner = SwiftSpinner.show(title, animated: false)
        spinner.outerCircle.strokeEnd = CGFloat(progress)
        return spinner
    }

    /// If set to true, hiding a spinner causes scheduled spinners to be canceled
    public static var hideCancelsScheduledSpinners = true

    /// Hide the spinner
    ///
    /// - Parameter completion: A closure called upon completion
    public class func hide(_ completion: (() -> Void)? = nil) {
        let spinner = SwiftSpinner.shared

        spinner.dismissing = true

        NotificationCenter.default.removeObserver(spinner)
        if hideCancelsScheduledSpinners {
            delayedTokens.removeAll()
        }

        DispatchQueue.main.async(execute: {
            spinner.clearTapHandler()

            if spinner.superview == nil {
                spinner.dismissing = false
                return
            }

            UIView.animate(withDuration: SwiftSpinner.standardAnimationDuration, delay: 0.0, options: .curveEaseOut, animations: {
                spinner.blurView.contentView.alpha = 0
                spinner.blurView.effect = nil
            }, completion: {_ in
                spinner.blurView.contentView.alpha = 1
                spinner.removeFromSuperview()
                spinner.titleLabel.text = nil
                spinner.dismissing = false

                completion?()
            })

            spinner.animating = false
        })
    }

    /// Set the default title font
    ///
    /// - Parameter font: The title font
    public class func setTitleFont(_ font: UIFont?) {
        let spinner = SwiftSpinner.shared

        spinner.currentTitleFont = font ?? spinner.defaultTitleFont
        spinner.titleLabel.font = font ?? spinner.defaultTitleFont
    }

    /// Set the default title color
    ///
    /// - Parameter color: The title color
    public class func setTitleColor(_ color: UIColor?) {
        let spinner = SwiftSpinner.shared

        spinner.titleLabel.textColor = color ?? spinner.defaultTitleColor
    }

    /// The spinner title
    public var title: String = "" {
        didSet {
            let spinner = SwiftSpinner.shared

            guard spinner.animating else {
                spinner.titleLabel.transform = CGAffineTransform.identity
                spinner.titleLabel.alpha = 1.0
                spinner.titleLabel.text = self.title
                return
            }

            UIView.animate(withDuration: 0.15, delay: 0.0, options: .curveEaseOut, animations: {
                spinner.titleLabel.transform = CGAffineTransform(scaleX: 0.75, y: 0.75)
                spinner.titleLabel.alpha = 0.2
                }, completion: { _ in
                    spinner.titleLabel.text = self.title
                    UIView.animate(withDuration: 0.35, delay: 0.0, usingSpringWithDamping: 0.35, initialSpringVelocity: 0.0, options: [], animations: {
                        spinner.titleLabel.transform = CGAffineTransform.identity
                        spinner.titleLabel.alpha = 1.0
                        }, completion: nil)
            })
        }
    }

    /// Observe the view frame and update the subviews layout
    public override var frame: CGRect {
        didSet {
            if frame == CGRect.zero {
                return
            }
            blurView.frame = bounds
            vibrancyView.frame = blurView.bounds
            titleLabel.center = vibrancyView.center
            outerCircleView.center = vibrancyView.center
            innerCircleView.center = vibrancyView.center
            layoutSubtitle()
        }
    }

    /// Start the spinning animation
    public var animating: Bool = false {

        willSet (shouldAnimate) {
            if shouldAnimate && !animating {
                spinInner()
                spinOuter()
            }
        }

        didSet {
            // Update UI
            if animating {
                self.outerCircle.strokeStart = 0.0
                self.outerCircle.strokeEnd = 0.45
                self.innerCircle.strokeStart = 0.5
                self.innerCircle.strokeEnd = 0.9
            } else {
                self.outerCircle.strokeStart = 0.0
                self.outerCircle.strokeEnd = 1.0
                self.innerCircle.strokeStart = 0.0
                self.innerCircle.strokeEnd = 1.0
            }
        }
    }

    /// Tap handler
    ///
    /// - Parameters:
    ///   - tap: The tap handler closure
    ///   - subtitleText: The optional subtitle
    public func addTapHandler(_ tap: @escaping (() -> Void), subtitle subtitleText: String? = nil) {
        clearTapHandler()

        //vibrancyView.contentView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: Selector("didTapSpinner")))
        tapHandler = tap

        if subtitleText != nil {
            subtitleLabel = UILabel()
            if let subtitle = subtitleLabel {
                subtitle.text = subtitleText
                subtitle.font = UIFont(name: self.currentTitleFont.familyName, size: currentTitleFont.pointSize * 0.8)
                subtitle.textColor = UIColor.white
                subtitle.numberOfLines = 0
                subtitle.textAlignment = .center
                subtitle.lineBreakMode = .byWordWrapping
                layoutSubtitle()
                vibrancyView.contentView.addSubview(subtitle)
            }
        }
    }

    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)

        if tapHandler != nil {
            tapHandler?()
            tapHandler = nil
        }
    }

    /// Remove the tap handler
    public func clearTapHandler() {
        isUserInteractionEnabled = false
        subtitleLabel?.removeFromSuperview()
        tapHandler = nil
    }

    // MARK: - Private interface

    //
    // Layout elements
    //

    private var blurEffectStyle: UIBlurEffect.Style = .dark
    private var blurEffect: UIBlurEffect!
    private var blurView: UIVisualEffectView!
    private var vibrancyView: UIVisualEffectView!

    private let defaultTitleFont = UIFont(name: "HelveticaNeue", size: 22.0)!
    private var currentTitleFont: UIFont

    private var defaultTitleColor = UIColor.white

    let frameSize = CGSize(width: 200.0, height: 200.0)

    private lazy var outerCircleView = UIView()
    private lazy var innerCircleView = UIView()

    private let outerCircle = CAShapeLayer()
    private let innerCircle = CAShapeLayer()

    required public init?(coder aDecoder: NSCoder) {
        fatalError("Not coder compliant")
    }

    private var currentOuterRotation: CGFloat = 0.0
    private var currentInnerRotation: CGFloat = 0.1

    private var dismissing: Bool = false

    private func spinOuter() {
        if superview == nil {
            return
        }

        let duration = Double(Float(arc4random()) /  Float(UInt32.max)) * 2.0 + 1.5
        let randomRotation = Double(Float(arc4random()) /  Float(UInt32.max)) * (Double.pi / 4) + (Double.pi / 4)

        //outer circle
        UIView.animate(withDuration: duration, delay: 0.0, usingSpringWithDamping: 0.4, initialSpringVelocity: 0.0, options: [], animations: {
            self.currentOuterRotation -= CGFloat(randomRotation)
            self.outerCircleView.transform = CGAffineTransform(rotationAngle: self.currentOuterRotation)
            }, completion: {_ in
                let waitDuration = Double(Float(arc4random()) /  Float(UInt32.max)) * 1.0 + 1.0
                self.delay(waitDuration, completion: {
                    if self.animating {
                        self.spinOuter()
                    }
                })
        })
    }

    private func spinInner() {
        if superview == nil {
            return
        }

        //inner circle
        UIView.animate(withDuration: 0.5, delay: 0.0, usingSpringWithDamping: 0.5, initialSpringVelocity: 0.0, options: [], animations: {
            self.currentInnerRotation += CGFloat(Double.pi / 4)
            self.innerCircleView.transform = CGAffineTransform(rotationAngle: self.currentInnerRotation)
            }, completion: { _ in
                self.delay(0.5, completion: {
                    if self.animating {
                        self.spinInner()
                    }
                })
        })
    }
    
    @objc func orientationChangedAction() {
        if let _ = SwiftSpinner.containerView() {
            SwiftSpinner.shared.setNeedsLayout()
        }
    }
    
    public func updateFrame() {
        if let containerView = SwiftSpinner.containerView() {
            SwiftSpinner.shared.frame = containerView.bounds
            containerView.bringSubviewToFront(SwiftSpinner.shared)
        }
    }

    // MARK: - Util methods

    func delay(_ seconds: Double, completion:@escaping () -> Void) {
        let popTime = DispatchTime.now() + Double(Int64( Double(NSEC_PER_SEC) * seconds )) / Double(NSEC_PER_SEC)

        DispatchQueue.main.asyncAfter(deadline: popTime) {
            completion()
        }
    }

    fileprivate func layoutSubtitle() {
        if let subtitle = subtitleLabel {
            subtitle.bounds.size = subtitle.sizeThatFits(bounds.insetBy(dx: 20.0, dy: 0.0).size)
            var safeArea: CGFloat = 0
            if #available(iOS 11.0, tvOS 11.0, *) {
                safeArea = superview?.safeAreaInsets.bottom ?? 0
            }
            subtitle.center = CGPoint(x: bounds.midX, y: bounds.maxY - subtitle.bounds.midY - subtitle.font.pointSize - safeArea)
        }
    }

    override public func layoutSubviews() {
        super.layoutSubviews()
        updateFrame()
    }

    // MARK: - Tap handler
    private var tapHandler: (() -> Void)?
    func didTapSpinner() {
        tapHandler?()
    }
}
